README.md
# Sensirion_SCD40_CO2_Sensor


extras/CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

 * [`added`]    SCD40: Initial support

## [2.0.2] - 2019-11-01

 * [`fixed`]    SCD30: Fix buffer overflow when reading measurements that was
                introduced in 2.0.1.

## [2.0.1] - 2019-10-17

 * [`fixed`]    Copy correct `AUTHORS`, `CHANGELOG.md`, `LICENSE`, and
                `README.md` files to target locations when running the `release`
                target of the driver's root Makefile.
 * [`removed`]  Remove the `AUTHORS` file from the driver and the
                `embedded-common` submodule, as it adds more noise than benefit.
                The contributors can be found in the git log.
 * [`changed`]  Avoid converting twice between endianness in
                `scd30_read_measurement`. (improves performance and portablity)
 * [`changed`]  Cleanup scd30 example

## [2.0.0] - 2019-10-01

 * [`changed`]  Functions are now dedicated per sensor (e.g. `scd30_probe()`
                instead of `scd_probe()`)
 * [`changed`]  Move the i2c init call out of `probe()` and into the example
 * [`added`]    Add `scd30_read_serial()` to read out the serial number
 * [`added`]    embedded-common: Add multi-bus support along with an example.
 * [`fixed`]    Fix compilation warnings when compiling the linux user space
                sample implementation with `CONFIG_I2C_TYPE` set to `sw_i2c`
 * [`changed`]  Split out `default_config.inc` from Makefile to configure paths
                and CFLAGS for the SCD30 driver
 * [`changed`]  Only one example with either `hw_i2c` or `sw_i2c` is build,
                depending on `CONFIG_I2C_TYPE`. Defaults to `hw_i2c`.
 * [`fixed`]    Fix clock stretching timeout that might occur when
                `CONFIG_I2C_TYPE` is set to `sw_i2c`

## [1.0.0] - 2019-05-14

 * `[fixed]` Add processing delay on i2c writes
 * `[fixed]` Potential issues regarding strict aliasing
 * `[changed]` Use stdint types (`uint16_t` instead of `u16`)

## [0.9.0] - 2018-10-08

 * `[fixed]` Measurement readout
 * `[added]` Self calibration
   * `scd_get_automatic_self_calibration`
   * `scd_set_forced_recalibration`
   * `scd_enable_automatic_self_calibration`
 * `[fixed]` Parameters on big-endian systems

## [0.1.0] - 2018-07-20

 * Initial support for SCD30

[Unreleased]: https://github.com/Sensirion/embedded-scd/compare/2.0.2...master
[2.0.2]: https://github.com/Sensirion/embedded-scd/compare/2.0.1...2.0.2
[2.0.1]: https://github.com/Sensirion/embedded-scd/compare/2.0.0...2.0.1
[2.0.0]: https://github.com/Sensirion/embedded-scd/compare/1.0.0...2.0.0
[1.0.0]: https://github.com/Sensirion/embedded-scd/compare/0.9.0...1.0.0
[0.9.0]: https://github.com/Sensirion/embedded-scd/compare/0.1.0...0.9.0
[0.1.0]: https://github.com/Sensirion/embedded-scd/releases/tag/0.1.0


src/scd40.c
/*
 * Copyright (c) 2019, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "scd40.h"
#include "scd_git_version.h"
#include "sensirion_arch_config.h"
#include "sensirion_common.h"
#include "sensirion_i2c.h"

#define SCD40_CMD_COMMIT_EEPROM 0x3615
#define SCD40_CMD_COMMIT_EEPROM_DURATION_US 6000000
#define SCD40_CMD_FACTORY_RESET 0x3632
#define SCD40_CMD_READ_AUTO_SELF_CALIBRATION 0x2313
#define SCD40_CMD_READ_MEASUREMENT 0xEC05
#define SCD40_CMD_READ_MEASUREMENT_NUM_TRIES 3
#define SCD40_CMD_READ_FEATURE_SET_VERSION 0x202f
#define SCD40_CMD_READ_SERIAL 0x3682
#define SCD40_CMD_RESET 0x3646
#define SCD40_CMD_SET_ALTITUDE 0x2427
#define SCD40_CMD_SET_AMBIENT_PRESSURE 0xE000
#define SCD40_CMD_SET_FORCED_RECALIBRATION 0x362F
#define SCD40_CMD_SET_TEMPERATURE_OFFSET 0x241D
#define SCD40_CMD_START_PERIODIC_MEASUREMENT_HIGH_PERFORMANCE 0x21B1
#define SCD40_CMD_START_PERIODIC_MEASUREMENT_LOW_POWER 0x21AC
#define SCD40_CMD_START_PERIODIC_MEASUREMENT_ULTRA_LOW_POWER 0x21A7
#define SCD40_CMD_STOP_PERIODIC_MEASUREMENT 0x3F86
#define SCD40_CMD_STOP_PERIODIC_MEASUREMENT_DURATION_USEC 30000
#define SCD40_CMD_SINGLE_SHOT_MEASUREMENT 0x2196
#define SCD40_CMD_SINGLE_SHOT_MEASUREMENT_DURATION_US 5000000  // 5 seconds
#define SCD40_CMD_WRITE_AUTO_SELF_CALIBRATION 0x2416

#define SCD40_CMD_SHORT_DURATION_US 10000

#define SCD40_SERIAL_NUM_WORDS 16

#define SCD40_MAX_BUFFER_WORDS 24
#define SCD40_CMD_SINGLE_WORD_BUF_LEN                                          \
    (SENSIRION_COMMAND_SIZE + SENSIRION_WORD_SIZE + CRC8_LEN)

static int16_t scd40_start_periodic_measurement_cmd(uint16_t cmd) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd_with_args(SCD40_I2C_ADDRESS, cmd, NULL, 0);
    if (ret)
        return ret;

    sensirion_sleep_usec(SCD40_CMD_SHORT_DURATION_US);
    return 0;
}

int16_t scd40_start_periodic_measurement() {
    return scd40_start_periodic_measurement_cmd(
        SCD40_CMD_START_PERIODIC_MEASUREMENT_HIGH_PERFORMANCE);
}

int16_t scd40_start_low_power_periodic_measurement() {
    return scd40_start_periodic_measurement_cmd(
        SCD40_CMD_START_PERIODIC_MEASUREMENT_LOW_POWER);
}

int16_t scd40_start_ultra_low_power() {
    return scd40_start_periodic_measurement_cmd(
        SCD40_CMD_START_PERIODIC_MEASUREMENT_ULTRA_LOW_POWER);
}

int16_t scd40_stop_periodic_measurement() {
    int16_t ret;

    ret = sensirion_i2c_write_cmd(SCD40_I2C_ADDRESS,
                                  SCD40_CMD_STOP_PERIODIC_MEASUREMENT);
    if (ret)
        return ret;

    sensirion_sleep_usec(SCD40_CMD_STOP_PERIODIC_MEASUREMENT_DURATION_USEC);
    return 0;
}

int16_t scd40_read_measurement(uint16_t *co2_ppm, int32_t *temperature,
                               int32_t *humidity) {
    uint16_t words[3];
    int16_t ret;
    int16_t tries;

    for (tries = 0; tries < SCD40_CMD_READ_MEASUREMENT_NUM_TRIES; ++tries) {
        ret = sensirion_i2c_read_words(SCD40_I2C_ADDRESS, words,
                                       SENSIRION_NUM_WORDS(words));
        if (ret == 0)
            break;

        sensirion_sleep_usec(100000); /* wait 100ms before retrying */
    }

    if (ret)
        return ret;

    *co2_ppm = words[0];
    /**
     * formulas for conversion of the sensor signals, optimized for fixed point
     * algebra:
     * Temperature = 175 * S_T / 2^16 - 45
     * Relative Humidity = 100 * S_RH / 2^16
     */
    *temperature = ((21875 * (int32_t)words[1]) >> 13) - 45000;
    *humidity = ((12500 * (int32_t)words[2]) >> 13);
    return 0;
}

int16_t scd40_measure_co2_temperature_and_humidity(void) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd(SCD40_I2C_ADDRESS,
                                  SCD40_CMD_SINGLE_SHOT_MEASUREMENT);
    if (ret)
        return ret;

    sensirion_sleep_usec(SCD40_CMD_SHORT_DURATION_US);
    return 0;
}

int16_t scd40_set_temperature_offset(int32_t temperature_offset) {
    int16_t ret;
    uint16_t offset;

    if (temperature_offset < 0 || temperature_offset > 174760)
        return SCD40_INVALID_VALUE;

    /* offset = temperature_offset * 2^16 / (175 * 1000) ->
     * 65536 / 175000 == 0.3745  ->  0.3745 * 2^3 = 2.996 */
    offset = (uint16_t)(temperature_offset >> 3) * 3;

    ret = sensirion_i2c_write_cmd_with_args(
        SCD40_I2C_ADDRESS, SCD40_CMD_SET_TEMPERATURE_OFFSET, &offset,
        SENSIRION_NUM_WORDS(offset));
    if (ret)
        return ret;

    sensirion_sleep_usec(SCD40_CMD_SHORT_DURATION_US);
    return 0;
}

int16_t scd40_set_altitude(uint16_t altitude_meters) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd_with_args(
        SCD40_I2C_ADDRESS, SCD40_CMD_SET_ALTITUDE, &altitude_meters,
        SENSIRION_NUM_WORDS(altitude_meters));
    if (ret)
        return ret;

    sensirion_sleep_usec(SCD40_CMD_SHORT_DURATION_US);
    return 0;
}

int16_t scd40_set_ambient_pressure(uint16_t pressure_pascal) {
    int16_t ret;
    const uint16_t sig_p = pressure_pascal * 100;

    if (pressure_pascal > 655)
        return SCD40_INVALID_VALUE;

    ret = sensirion_i2c_write_cmd_with_args(SCD40_I2C_ADDRESS,
                                            SCD40_CMD_SET_AMBIENT_PRESSURE,
                                            &sig_p, SENSIRION_NUM_WORDS(sig_p));
    if (ret)
        return ret;

    sensirion_sleep_usec(SCD40_CMD_SHORT_DURATION_US);
    return 0;
}

int16_t scd40_get_automatic_self_calibration(uint8_t *asc_enabled) {
    uint16_t word;
    int16_t ret;

    ret = sensirion_i2c_delayed_read_cmd(
        SCD40_I2C_ADDRESS, SCD40_CMD_READ_AUTO_SELF_CALIBRATION,
        SCD40_CMD_SHORT_DURATION_US, &word, SENSIRION_NUM_WORDS(word));
    if (ret != STATUS_OK)
        return ret;

    *asc_enabled = (uint8_t)word;

    return STATUS_OK;
}

int16_t scd40_enable_automatic_self_calibration(uint8_t enable_asc) {
    int16_t ret;
    uint16_t asc = !!enable_asc;

    ret = sensirion_i2c_write_cmd_with_args(
        SCD40_I2C_ADDRESS, SCD40_CMD_WRITE_AUTO_SELF_CALIBRATION, &asc,
        SENSIRION_NUM_WORDS(asc));
    sensirion_sleep_usec(SCD40_CMD_SHORT_DURATION_US);

    return ret;
}

int16_t scd40_set_forced_recalibration(uint16_t co2_ppm) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd_with_args(
        SCD40_I2C_ADDRESS, SCD40_CMD_SET_FORCED_RECALIBRATION, &co2_ppm,
        SENSIRION_NUM_WORDS(co2_ppm));
    sensirion_sleep_usec(SCD40_CMD_SHORT_DURATION_US);

    return ret;
}

int16_t scd40_read_serial(uint64_t *serial) {
    int16_t ret;
    uint16_t words[3];

    ret = sensirion_i2c_delayed_read_cmd(
        SCD40_I2C_ADDRESS, SCD40_CMD_READ_SERIAL, SCD40_CMD_SHORT_DURATION_US,
        words, SENSIRION_NUM_WORDS(words));
    if (ret)
        return ret;

    *serial = (((uint64_t)words[0] << 4) | ((uint64_t)words[1] << 2) |
               (uint64_t)words[2]);
    return 0;
}

int16_t scd40_probe(void) {
    /* try to stop a pending measurement */
    return scd40_stop_periodic_measurement();
}

int16_t scd40_read_feature_set_version(uint16_t *feature_set) {
    return sensirion_i2c_delayed_read_cmd(
        SCD40_I2C_ADDRESS, SCD40_CMD_READ_FEATURE_SET_VERSION,
        SCD40_CMD_SHORT_DURATION_US, feature_set,
        SENSIRION_NUM_WORDS(*feature_set));
}

int16_t scd40_factory_reset(void) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd(SCD40_I2C_ADDRESS, SCD40_CMD_FACTORY_RESET);
    if (ret)
        return ret;

    sensirion_sleep_usec(SCD40_CMD_COMMIT_EEPROM_DURATION_US);
    return 0;
}

int16_t scd40_reset(void) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd(SCD40_I2C_ADDRESS, SCD40_CMD_RESET);
    if (ret)
        return ret;

    sensirion_sleep_usec(SCD40_CMD_SHORT_DURATION_US);
    return 0;
}

int16_t scd40_persist_settings(void) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd(SCD40_I2C_ADDRESS, SCD40_CMD_COMMIT_EEPROM);
    if (ret)
        return ret;

    sensirion_sleep_usec(SCD40_CMD_COMMIT_EEPROM_DURATION_US);
    return 0;
}


extras/scd40_example_usage.c
/*
 * Copyright (c) 2019, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "scd40.h"

#include <stdio.h> /* printf */

/* TO USE CONSOLE OUTPUT (printf) YOU MAY NEED TO ADAPT THE
 * INCLUDE ABOVE OR DEFINE IT ACCORDING TO YOUR PLATFORM.
 * #define printf(...)
 */

int main(void) {
    uint16_t co2_ppm;
    int32_t temperature, relative_humidity;
    int16_t err;
    uint16_t interval_in_seconds = 2;

    /* Initialize I2C */
    sensirion_i2c_init();

    /* Busy loop for initialization, because the main loop does not work without
     * a sensor.
     */
    while (scd40_probe() != STATUS_OK) {
        printf("SCD40 sensor probing failed\n");
        sensirion_sleep_usec(1000000u);
    }
    printf("SCD40 sensor probing successful\n");

    sensirion_sleep_usec(20000u);
    scd40_start_periodic_measurement();
    sensirion_sleep_usec(interval_in_seconds * 1000000u);

    while (1) {
        /* Measure co2, temperature and relative humidity and store into
         * variables.
         */
        err =
            scd40_read_measurement(&co2_ppm, &temperature, &relative_humidity);
        if (err != STATUS_OK) {
            printf("error reading measurement\n");

        } else {
            printf("measured co2 concentration: %u ppm, "
                   "measured temperature: %0.3f degreeCelsius, "
                   "measured humidity: %0.3f %%RH\n",
                   co2_ppm, temperature / 1000.0f, relative_humidity / 1000.0f);
        }

        sensirion_sleep_usec(interval_in_seconds * 1000000u);
    }

    scd40_stop_periodic_measurement();

    sensirion_i2c_release();
    return 0;
}


src/scd40.h
/*
 * Copyright (c) 2019, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SCD40_H
#define SCD40_H
#include "sensirion_arch_config.h"
#include "sensirion_common.h"
#include "sensirion_i2c.h"

#ifdef __cplusplus
extern "C" {
#endif

#define SCD40_I2C_ADDRESS 0x62

#define SCD40_INVALID_VALUE (-(1 << 14))

#define SCD40_FEATURE_SET_MAJOR_VERSION(fs) (((fs)&0xE0) >> 5)
#define SCD40_FEATURE_SET_MINOR_VERSION(fs) ((fs)&0x1F)

/**
 * scd40_probe() - check if the SCD sensor is available and initialize it
 *
 * Note that as part of the initialization, any ongoing measurements are stopped
 * with `scd40_stop_periodic_measurement()` since the sensor does not accept
 * commands when a periodic measurement is running.
 *
 * @return  0 on success, an error code otherwise.
 */
int16_t scd40_probe(void);

/**
 * scd40_start_periodic_measurement() - Start continuous measurement to measure
 * CO2 concentration in the highest possible accuracy, relative humidity and
 * temperature.
 *
 * Measurement data which is not read from the sensor is continuously
 * overwritten.
 * The continuous measurement status is saved in non-volatile memory. The last
 * measurement mode is resumed after repowering.
 *
 * @return  0 if the command was successful, an error code otherwise
 */
int16_t scd40_start_periodic_measurement(void);

/**
 * scd40_start_low_power_periodic_measurement - Start measurement in low power
 * mode.
 *
 * @return  0 if the command was successful, an error code otherwise
 */
int16_t scd40_start_low_power_periodic_measurement(void);

/**
 * scd40_start_ultra_low_power_periodic_measurement - Start measurement in
 * ultra low power mode.
 *
 * @return  0 if the command was successful, an error code otherwise
 */
int16_t scd40_start_ultra_low_power_periodic_measurement(void);

/**
 * scd40_stop_periodic_measurement() - Stop the continuous measurement
 *
 * @return  0 if the command was successful, an error code otherwise
 */
int16_t scd40_stop_periodic_measurement(void);

/**
 * scd40_read_measurement() - Read out the last measurement from an ongoing
 * periodic measurement or from a single shot measurement
 *
 * Temperature is returned in [degree Celsius], multiplied by 1000,
 * and relative humidity in [percent relative humidity], multiplied by 1000.
 *
 * @param co2_ppm       CO2 concentration in ppm
 * @param temperature   the address for the result of the temperature
 *                      measurement
 * @param humidity      the address for the result of the relative humidity
 *                      measurement
 *
 * @return              0 if the command was successful, an error code otherwise
 */
int16_t scd40_read_measurement(uint16_t *co2_ppm, int32_t *temperature,
                               int32_t *humidity);

/**
 * scd40_measure_co2_temperature_and_humidity - Start a single-shot measurement
 *
 * A single-shot measurement is available after
 * SCD40_READ_CO2_TEMPERATURE_AND_HUMIDITY_DURATION_USEC micoseconds
 * with the method `scd40_read_measurement`
 *
 * @return              0 if the command was successful, an error code otherwise
 */
int16_t scd40_measure_co2_temperature_and_humidity(void);

/**
 * scd40_set_temperature_offset() - Set the temperature offset
 *
 * The on-board RH/T sensor is influenced by thermal self-heating of SCD40 and
 * other electrical components. Design-in alters the thermal properties of SCD40
 * such that temperature and humidity offsets may occur when operating the
 * sensor in end-customer devices. Compensation of those effects is achievable
 * by writing the temperature offset found in continuous operation of the device
 * into the sensor.
 *
 * @param temperature_offset    Temperature offset in [degree Celsius],
 *                              multiplied by 1000, i.e. one tick corresponds to
 *                              0.001 degrees C.
 *                              Only positive values and values < 175C are
 *                              permissible, otherwise SCD40_INVALID_VALUE is
 *                              returned.
 *
 * @return                      0 if the command was successful, an error code
 *                              otherwise
 */
int16_t scd40_set_temperature_offset(int32_t temperature_offset);

/**
 * scd40_set_altitude() - Set the altitude above sea level
 *
 * Measurements of CO2 concentration are influenced by altitude. When a value is
 * set, the altitude-effect is compensated. The altitude setting is disregarded
 * when an ambient pressure is set on the sensor with
 * scd40_start_periodic_measurement.
 * The altitude is saved in non-volatile memory. The last set value will be used
 * after repowering.
 *
 * @param altitude  altitude in meters above sea level, 0 meters is the default
 *                  value and disables altitude compensation
 *
 * @return          0 if the command was successful, an error code otherwise
 */
int16_t scd40_set_altitude(uint16_t altitude_meters);

/**
 * scd40_set_ambient_pressure() - Set the ambient pressure
 *
 * Measurements of CO2 concentration are influenced by pressure. When a value is
 * set, the pressure-effect is compensated. The altitude setting is disregarded
 * when an ambient pressure is set on the sensor.
 *
 * @param pressure_pascal   The ambient pressure in Pascal
 *
 * @return              0 if the command was successful, an error code otherwise
 */
int16_t scd40_set_ambient_pressure(uint16_t pressure_pascal);

/**
 * scd40_get_automatic_self_calibration() - Read if the sensor's automatic self
 * calibration is enabled or disabled
 *
 * See scd40_enable_automatic_self_calibration() for more details.
 *
 * @param asc_enabled   Pointer to memory of where to set the self calibration
 *                      state. 1 if ASC is enabled, 0 if ASC disabled. Remains
 *                      untouched if return is non-zero.
 *
 * @return              0 if the command was successful, an error code otherwise
 */
int16_t scd40_get_automatic_self_calibration(uint8_t *asc_enabled);

/**
 * scd40_enable_automatic_self_calibration() - Enable or disable the sensor's
 * automatic self calibration
 *
 * When activated for the first time a period of minimum 7 days is needed so
 * that the algorithm can find its initial parameter set for ASC.
 * The sensor has to be exposed to fresh air for at least 1 hour every day.
 * Refer to the datasheet for further conditions
 *
 * ASC status is saved in non-volatile memory. When the sensor is powered down
 * while ASC is activated SCD40 will continue with automatic self-calibration
 * after repowering without sending the command.
 *
 * @param enable_asc    enable ASC if non-zero, disable otherwise
 *
 * @return              0 if the command was successful, an error code otherwise
 */
int16_t scd40_enable_automatic_self_calibration(uint8_t enable_asc);

/**
 * scd40_set_forced_recalibration() - Forcibly recalibrate the sensor to a known
 * value.
 *
 * Forced recalibration (FRC) is used to compensate for sensor drifts when a
 * reference value of the CO2 concentration in close proximity to the SCD40 is
 * available.
 *
 * For best results the sensor has to be run in a stable environment in
 * continuous mode at a measurement rate of 2s for at least two minutes before
 * applying the calibration command and sending the reference value.
 * Setting a reference CO2 concentration will overwrite the settings from ASC
 * (see scd40_enable_automatic_self_calibration) and vice-versa. The reference
 * CO2 concentration has to be in the range 400..2000 ppm.
 *
 * FRC value is saved in non-volatile memory, the last set FRC value will be
 * used for field-calibration after repowering.
 *
 * @param co2_ppm   recalibrate to this specific co2 concentration
 *
 * @return          0 if the command was successful, an error code otherwise
 */
int16_t scd40_set_forced_recalibration(uint16_t co2_ppm);

/**
 * scd40_read_serial() - Read out the serial number
 *
 * Note that only the lower 48 bits are returned from the sensor, the upper bits
 * are always 0.
 *
 * @param serial    the address for the result of the serial number.
 * @return          0 if the command was successful, else an error code.
 */
int16_t scd40_read_serial(uint64_t *serial);

/**
 * scd40_read_feature_set_version() - Read the feature set version
 *
 * @param feature_set   the address for the result of the feature set version.
 * @return              0 on success, an error code otherwise.
 */
int16_t scd40_read_feature_set_version(uint16_t *feature_set);

/**
 * scd40_factory_reset - Reset all settings to factory default
 *
 * @return  0 on success, an error code otherwise.
 */
int16_t scd40_factory_reset(void);

/**
 * scd40_reset - Reset (aka restart or reinit) the sensor
 *
 * Imitates a soft reset of the sensor. Before sending the command, the stop
 * measurement command must be issued.
 *
 * @return  0 on success, an error code otherwise.
 */
int16_t scd40_reset(void);

/**
 * scd40_persist_settings - Commit all settings to the sensor's internal
 * persistance
 *
 * Configurations on the SCD40 are by default stored in the volatile memory only
 * and will be lost after a power-cycle. This command stores current
 * configurations in the EEPROM of the SCD40, making them resistant to
 * power-cycling.
 * Note that the command should only be sent after all configuration are set to
 * minimize the number of write/erase cycles on the EEPROM.
 *
 * @return  0 on success, an error code otherwise.
 */
int16_t scd40_persist_settings(void);

#ifdef __cplusplus
}
#endif

#endif /* SCD40_H */
